### PR TITLE
Federation Averaging with Differential Privacy

### PR DESCRIPTION
**Pull request by BGluth**
So this is an initial impl. It's definitely not fully correct (but it runs), and there are still some major TODOs. I've added an example of running this in `Using the Runner.ipynb`.

Noteable TODOs:

- Theta is not being calculated correctly in `user_update_fed_avg`.
- _get_random_selection_of_user_idxs is always picking a 1/5th sample of users each time (it probably should be variable).
- Moments accountant stuff is still all a placeholder.
- Not heavily tested!!

### CODE REVIEW

**Comment by crankycoder** [comment]
@maharshmellow @zhaoyuxuan  let's get some testcases put into this code so that everyone can review the codebase for understanding and to isolate some early bugs.

**Comment by BGluth** [comment]
@crankycoder Ok I think this is ready for review as it now converges on the blob dataset.

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py]: I would rename this parameter to something more reflective of its purpose, such as `user_weight_cap`

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py]: The initialization of the model weights is not related to the moments accountant. I would rename this function to drop the reference.

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py]: We only really need a single `theta` variable which gets updated in each round. I suggest dropping `theta_0` and `prev_theta`

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py]: The selection probability should be a param read from `prms`. It determines the proportion of clients to include in each round, anlagous to `client_fraction` in `simulation_util.server_update`.

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py]: It's constant because it's related to the maximum theoretical contribution of any single client, which is determined by S (`prms.sensitivity`), as opposed to observed contributions.

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py]: This seems to restrict the user's data to a subset. This is not necessary.

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py:72]: This should use `theta`, the current weights, not `theta_0`.

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py]: As above, don't need to maintain `prev_theta` as a separate variable.

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py]: It would be better to use a conditional form, to avoid dividing if the norm is small. Ie:
```python
norm = np.linalg.norm(vec)
if norm > sensitivity:
  vec *= sensitivity / norm
```

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py]: I don't think there's any reason to drop the last batch, even if it's smaller than the target size.

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py]: This should use `user_sel_prob`.

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py:217]: As mentioned above, drop the "moment accountant" reference from the function name.

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py]: In the paper, n_k is the number of training examples for user k. Ie., user's weight is the ratio of the size of their personal dataset to the cap (`weight_mod`), bounded at 1.

I would restrict this function to computing the weight for a single user, given their dataset, and move the rest to the main runner function above. The runner function would then loop over users computing individual user weights and the sum.

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py]: It looks like this limits the data used for a given user in the current round to a random subset of size at least as large as the batch size. This is not necessary - each round should cover the full user data, split into batches (as is done in `user_update_fed_avg`).

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [inline mozfldp/fed_avg_w_dp.py]: Cf comment for `_get_data_for_user_for_round`: this function shouldn't be necessary.

**Review by dzeber — CHANGES_REQUESTED** [review]
Good work! This was a challenging implementation covering a whole bunch of unfamiliar concepts. There are a few points that need to be ironed out, commented inline.

A couple of other general points:
- Make sure to document the code. There are some comments scattered through, but each function should have a docstring, at the very least a one-liner summarizing what it does.
- Rather than using a class just to wrap the parameters, it would be better to have the majority of these functions as methods in there as well, so that they get internal access to the params, rather than still having to pass the param object around. However, let's keep that as a separate issue so as not to block this PR. Also it will depend on the ongoing architecture change to an actual client-server model.
- I don't see a benefit to keep converting between the (coef, intercept) and theta forms for the model weights. Since the `SGDClassifier` uses (coef, intercept), let's just use this throughout. However, let's also break this off into a separate issue after the PR is merged.

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: I'm a bit confused on this one. Like we're just pulling random entries from a user/round right?

Are you thinking about ignoring the grouping of data/user and instead just pulling data randomly from the entire set?

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: Fixed.

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py:217]: Done.

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: Changed.

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: Changed.

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: :+1:

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: Done.

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: Ok I changed it so all of the user's data is used when they are randomly picked.

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py:72]: Done.

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: Changed.

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: Optimized!

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: Agreed. Fixed.

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: This is now redone. I think you may want to give it a second glance however.

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: Done.

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: Removed.

File: mozfldp/fed_avg_w_dp.py
Comment (BGluth) [reply mozfldp/fed_avg_w_dp.py]: Ok I made the fixes for this.

**Comment by BGluth** [comment]
Ok @dzeber. I made all the requested changes. Ball is back in your court.

File: mozfldp/fed_avg_w_dp.py
Comment (dzeber) [reply mozfldp/fed_avg_w_dp.py]: I think your updated version is fine. We could optimize this using numpy vectorization but the logic will be the same.

**Review by dzeber — APPROVED** [review]

**Comment by dzeber** [comment]
Your fixes look good. I think this is ready to go.

**Pull request merged**
